miio/gateway.py

```diff
@@ -536,38 +513,216 @@ class SubDevice:
+        return self._battery
 
-    def get_firmware_version(self) -> Optional[int]:
-        """Returns firmware version"""
+    @command()
+    def subdevice_send(self, command):
+        """Send a command/query to the subdevice"""
+        try:
+            return self.gw.send(command, [self.sid])
+        except Exception as ex:
+            _LOGGER.error(
+                "Got an exception while sending command %s: %s", command, ex, exc_info=True
+            )
+            return None
+
+    @command()
+    def subdevice_send_arg(self, command, arguments):
+        """Send a command/query including arguments to the subdevice"""
         try:
-            return self.gw.send("get_device_prop", [self.sid, "fw_ver"])[0]
+            return self.gw.send(command, arguments, extra_parameters={"sid": self.sid})
+        except Exception as ex:
+            _LOGGER.error(
+                "Got an exception while sending command '%s' with arguments '%s': %s", command, str(arguments), ex, exc_info=True
+            )
+            return None
+
+    @command(click.argument("property"))
+    def get_subdevice_prop(self, property):
+        """Get the value of a property of the subdevice."""
+        try:
+            return self.gw.send("get_device_prop", [self.sid, property])
         except Exception as ex:
             _LOGGER.debug(
-                "Got an exception while fetching fw_ver: %s", ex, exc_info=True
+                "Got an exception while fetching property %s: %s", property, ex, exc_info=True
             )
             return None
 
-    def __repr__(self):
-        return "<Subdevice %s: %s fw: %s bat: %s>" % (
-            self.type,
-            self.sid,
-            self.get_firmware_version(),
-            self.battery(),
-        )
+    @command(click.argument("properties", nargs=-1))
+    def get_subdevice_prop_exp(self, properties):
+        """Get the value of a bunch of properties of the subdevice."""
+        try:
+            return self.gw.send("get_device_prop_exp", [[self.sid] + list(properties)]).pop()
+        except Exception as ex:
+            _LOGGER.debug(
+                "Got an exception while fetching properties %s: %s", properties, ex, exc_info=True
+            )
+            return None
+
+    @command(click.argument("property"), click.argument("value"))
+    def set_subdevice_prop(self, property, value):
+        """Set a device property of the subdevice."""
+        try:
+            return self.gw.send("set_device_prop", {"sid": self.sid, property: value})
+        except Exception as ex:
+            _LOGGER.debug(
+                "Got an exception while setting propertie %s to value %s: %s", property, str(value), ex, exc_info=True
+            )
+            return None
+
+    @command()
+    def unpair(self):
+        return self.subdevice_send("remove_device")
+
+    @command()
+    def get_battery(self):
+        self._battery = self.subdevice_send("get_battery").pop()
```

Comment:
Is there some specific reason for caching this?

Reply:
My idea was that this function can be called in an update command of the device_specific class that updates all properties in one command.
The HomeAssistant properties can then be directly mapped on the properties of this class.
So in case of the AqaraTH sensor you call update once and it updates pressure, temperature, humidity and battery level in one go.